Frontend log bundle

sja opened the pull request:
As announced at [Eclipse Forum](https://www.eclipse.org/forums/index.php?t=msg&th=1072176&goto=1714706&#msg_1714706), I've implemented a UI-Log receiver.

I added a README which is normal Markdown but also a valid [API Blueprint](https://apiblueprint.org/) document, which can be used in e.g. [Apiary](https://apiary.io).

The PR for PaperUI will come later on.

Signed-off-by: Sebastian Janzen sebastian.janzen@innoq.com

sja commented:
Shouldn't this be merged?

maggu2810 commented on extensions/ui/org.eclipse.smarthome.ui.log/src/main/java/org/eclipse/smarthome/ui/log/internal/LogHandler.java:
```diff
@@ -0,0 +1,129 @@
+@Produces(MediaType.APPLICATION_JSON)
+public class LogHandler implements org.eclipse.smarthome.io.rest.RESTResource {
+
+    private final static Logger logger = LoggerFactory.getLogger(LogHandler.class);
```
Can you remove the static keyword?

maggu2810 commented on extensions/ui/org.eclipse.smarthome.ui.log/src/main/java/org/eclipse/smarthome/ui/log/internal/LogHandler.java:
```diff
@@ -0,0 +1,129 @@
+     * Rolling array to store the last LOG_BUFFER_LIMIT messages. Those can be fetched e.g. by a
+     * diagnostic UI to display errors of other clients, where e.g. the logs are not easily accessible.
+     */
+    private final static ArrayDeque<LogMessage> LOG_BUFFER = new ArrayDeque<>(LogConstants.LOG_BUFFER_LIMIT);
```
You can remove the static here, too

maggu2810 commented:
I added my comments, but looks good to me. Very good documented stuff. :+1: 

If there aren't any reasons not to do so, I think it could be merged.

sja commented:
I removed the statics and squashed.
Whats the background of this? I usually use a static Logger. Do you have a forum link or something?

maggu2810 commented:
* http://www.slf4j.org/faq.html#declared_static
* http://wiki.apache.org/commons/Logging/StaticLog

kaikreuzer commented on extensions/ui/org.eclipse.smarthome.ui.log/src/main/java/org/eclipse/smarthome/ui/log/internal/LogHandler.java:
```diff
@@ -0,0 +1,129 @@
+package org.eclipse.smarthome.ui.log.internal;
```
please add the license header

kaikreuzer commented on extensions/ui/org.eclipse.smarthome.ui.log/src/main/java/org/eclipse/smarthome/ui/log/internal/LogHandler.java:
```diff
@@ -0,0 +1,129 @@
+
+@Path("/log")
+@Produces(MediaType.APPLICATION_JSON)
+public class LogHandler implements org.eclipse.smarthome.io.rest.RESTResource {
```
why don't you import this class instead of using the fqcn?

kaikreuzer commented:
> The PR for PaperUI will come later on.

I was still waiting for this :-) Wouldn't it make sense to directly use this feature in the Paper UI, so that there is a reference implementation and it is actively used in ESH?

Please also add the bundle to the runtime.ui feature, so that it is included in the p2 repo.

kaikreuzer commented on extensions/ui/org.eclipse.smarthome.ui.log/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,16 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: 
+ io.swagger.annotations,
```
please make this optional as it is a compile time only dependency, right?

kaikreuzer commented:
Question: As the only functionality in here is a rest endpoint, shouldn't it be better be io.rest.log instead of ui.log?

sja commented:
Ok, will fix that as well.
I added it to ui.log because it's strongly related to UIs. But maybe you're right and I have to move that.

Yes, PaperUI would be a good reference. But it's useful for commercial UIs, too. So it's better to contribute parts instead of waiting until I have time to add it to PaperUI. Right?

cdjackson commented:
Cool - any chance that the GET API can add a filter to select specific bindings?

sja commented:
Hi Chris! Do you mean binding or bundle? How should I distinguish between them? All I POST is the severity and an URL. Of course, you can post an URI which has a the information encoded. Then we can later add a filter for the URIs to the GET method.
If you have UI bundles in mind, a UI can be a mashup of different UIs. Then it may be complicated to find the root of a message internally.

maggu2810 commented:
Perhaps it has not been clear, that the GET API only offers log messages that has been added using the POST request. This is nothing to get the log (e.g. slf4j log) that has been written by a specific bundle.

kaikreuzer commented:
> So it's better to contribute parts instead of waiting until I have time to add it to PaperUI. Right?

If we really do not use it ourselves anywhere, it would be good to have it included in the documentation somewhere (e.g. under https://github.com/eclipse/smarthome/tree/master/docs/documentation/features/ui).

But i the missing parts are not too much, I prefer having "complete" PRs, where nothing is left open, because the tendency then is that such a follow-up contribution will never happen. I would assume that it is not too hard to make the Paper UI post its exceptions to this new endpoint, so it would be great to have that in place.

cdjackson commented:
Hi,
Maybe I misunderstood the code (I didn't read it other than the comments :) ). I thought that this returned 'all' logged entries, but looking again it seem this GET only return entries logged through this POST - not from the logger.

I was ultimately thinking about being able to get log entries from some bundles - I have a nice log display for ZWave in HABmin that reads the log files and I was hoping I might be able to hook in to this for 'real-time' display. However, it's not doing quite what I thought, so disregard this (sorry).

Cheers
Chris

sja commented:
Ok, I moved the package to io.rest.logging, added it to the feature (hopefully right?). The follow-up PR whith the PaperUI extension is already pushed to my fork, but not ready yet. Will continue with this in 1-2 weeks.
(Just writing that to not make you think that I forgot this PR)

afuechsel commented on bundles/io/org.eclipse.smarthome.io.rest.log/src/main/java/org/eclipse/smarthome/io/rest/log/internal/LogHandler.java:
```diff
@@ -0,0 +1,137 @@
+     * Rolling array to store the last LOG_BUFFER_LIMIT messages. Those can be fetched e.g. by a
+     * diagnostic UI to display errors of other clients, where e.g. the logs are not easily accessible.
+     */
+    private final ArrayDeque<LogMessage> LOG_BUFFER = new ArrayDeque<>(LogConstants.LOG_BUFFER_LIMIT);
```
ArrayDeque is not thread save. If multiple threads are logging, you would risk a ConcurrentModificationException. Please change it to e.g. ConcurrentLinkedDeque.

afuechsel commented on bundles/io/org.eclipse.smarthome.io.rest.log/README.md:
```diff
@@ -0,0 +1,119 @@
+## Log levels [/rest/log/levels]
+
+### Get enabled log levels [GET]
+
```
Wouldn't it be good to offer also a PUT to change the log level via REST call?

kaikreuzer replied:
I think this feature is not about allowing to configure your server logging, but to allow UIs to log messages to the server side log. So GET should be fine.

sja commented:
I added ConcurrentLinkedDeque, currently blocked by [this issue](https://github.com/openhab/openhab2-addons/issues/1152)

kaikreuzer commented:
Hey @sja, just looking at the list of open PRs and found this one - sorry for having ignored it for a long while...

As far as I can see, it is ready to be merged, right? Could you please rebase to resolve the current merge conflicts? Once it is merged, is your follow-up PR ready to adapt the Paper UI to make use of this feature?

sja commented:
Ready for take-off?

kaikreuzer commented:
Thanks! By now, the license headers are a bit outdated, could you re-run "mvn license:format"?

@maggu2810 Where would we want to add this bundle to the Karaf features? Add it to the base-esh or as a separate feature?

sja commented:
Sorry, I'm a little out of practice.

But travis had a timeout here, it was not my fault.

maggu2810 commented:
> @maggu2810 Where would we want to add this bundle to the Karaf features? Add it to the base-esh or as a separate feature?

esh-rest-log

I can do it after this one has been merged.

maggu2810 commented:
> But Travis had a timeout here, it was not my fault.

It is not your fault. I agree.
But it is not correct that Travis timed out.

You should enlarge the tail stuff at the end.

```text
-------------------------------------------------------

 T E S T S

-------------------------------------------------------

Running org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest

Tests run: 16, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 17.783 sec <<< FAILURE! - in org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest

verify the HTTP status is handled correctly when it is not OK_200(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 1.245 sec  <<< FAILURE!

java.lang.AssertionError: Something went wrong with the removal of the registered Items

Expected: is <true>

     but: was <false>

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)

	at org.junit.Assert.assertThat(Assert.java:865)

	at org.junit.Assert$assertThat$0.callStatic(Unknown Source)

	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:53)

	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:157)

	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:173)

	at org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest.clearTheItemRegistry(FSInternetRadioHandlerOSGiTest.groovy:678)

	at org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest.tearDown(FSInternetRadioHandlerOSGiTest.groovy:127)

verify the preset channel is updated(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.42 sec

verify not registered RadioHandler when the thingTypeUID is not the DEFAULT_THING_TYPE_UID(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.022 sec

verify the volume is updated through the CHANNEL_VOLUME_PERCENT(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 3.107 sec

verify the mute channel is updated(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 2.059 sec

verify ONLINE status of a Thing with complete configuration(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.071 sec

verify the mode channel is updated(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 2.079 sec

verify OFFLINE Thing-status and NULL Item-state when the IP is NULL(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.328 sec

verify OFFLINE Thing-status and NULL Item-state when the PIN is wrong(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.394 sec

verify the volume is updated through the CHANNEL_VOLUME_ABSOLUTE(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 2.443 sec

verify OFFLINE Thing-status and NULL Item-state when the HTTP response cannot be parsed correctly(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.38 sec

verify OFFLINE Thing-status and NULL Item-state when the PORT is zero(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.31 sec

verify the power channel is updated(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 2.045 sec

verify the playInfoText channel is updated(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 1.033 sec

verify the playInfoName channel is updated(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 1.033 sec

verify OFFLINE Thing-status and NULL Item-state when the PIN is empty String(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioHandlerOSGiTest)  Time elapsed: 0.33 sec

Running org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioDiscoveryParticipantOSGiTest

Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.055 sec - in org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioDiscoveryParticipantOSGiTest

verify valid DiscoveryResult with FSInterntRadio device without base URL(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioDiscoveryParticipantOSGiTest)  Time elapsed: 0.034 sec

verify no discovery result for unknown device(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioDiscoveryParticipantOSGiTest)  Time elapsed: 0.001 sec

verify no discovery result for FSInternetRadio device with null details(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioDiscoveryParticipantOSGiTest)  Time elapsed: 0 sec

verify valid DiscoveryResult with completeFSInterntRadioDevice(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioDiscoveryParticipantOSGiTest)  Time elapsed: 0.001 sec

verify correct supported types(org.eclipse.smarthome.binding.fsinternetradio.test.FSInternetRadioDiscoveryParticipantOSGiTest)  Time elapsed: 0.004 sec

Results :

Failed tests: 

  FSInternetRadioHandlerOSGiTest.tearDown:127->clearTheItemRegistry:678 Something went wrong with the removal of the registered Items

Expected: is <true>

     but: was <false>

Tests run: 21, Failures: 1, Errors: 0, Skipped: 0
```
